### ISSUE TITLE
Plugin is not completely compatible with gradle 7.2

### ISSUE THREAD

**Issue by patton73** [johnrengelman/shadow#713]
### Shadow Version
7.0.0

### Gradle Version
7.2

After compilation the plugin prints those warning/errors.

Task :shadowJar
Execution optimizations have been disabled for task ':shadowJar' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '...x.jar'. Reason: Task ':distTar' uses this output of task ':shadowJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '...x.jar'. Reason: Task ':distZip' uses this output of task ':shadowJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '...x.jar'. Reason: Task ':startScripts' uses this output of task ':shadowJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.

> Task :startShadowScripts
Execution optimizations have been disabled for task ':startShadowScripts' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '...x.jar'. Reason: Task ':startShadowScripts' uses this output of task ':jar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.

> Task :shadowDistTar
Execution optimizations have been disabled for task ':shadowDistTar' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '...x.jar. Reason: Task ':shadowDistTar' uses this output of task ':jar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.

> Task :shadowDistZip
Execution optimizations have been disabled for task ':shadowDistZip' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '...x.jar'. Reason: Task ':shadowDistZip' uses this output of task ':jar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.

Is there a way to fix this behaviour?
Thanks in advance.

**Comment by jasonculverhouse** [comment]
Is there any way to get the code in master released? I think this fixes most of these compatibility issues. 

Also, I was unable to get this plugin to work properly if i publish to a local maven repo.   Are there any instructions to get that working correctly?

**Comment by vyazelenko** [comment]
@johnrengelman We are also awaiting a new release to be able to test with JDK 18-ea.

**Comment by patton73** [comment]
Is this project mantained? I see few commits latest months.

**Comment by johnrengelman** [comment]
7.1.0 is now out. 
sorry for the delays.

**Comment by patton73** [comment]
Unfortunately this does not resolve the warnings.

**Comment by johnrengelman** [comment]
@patton73 I don't see any deprecations in any of the automated tests or when I spin up a skeleton project.
```
$  gradle --version

------------------------------------------------------------
Gradle 7.2
------------------------------------------------------------

Build time:   2021-08-17 09:59:03 UTC
Revision:     a773786b58bb28710e3dc96c4d1a7063628952ad

Kotlin:       1.5.21
Groovy:       3.0.8
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          1.8.0_121 (Oracle Corporation 25.121-b13)
OS:           Mac OS X 10.16 x86_64

$  gradle assembleShadowDist

BUILD SUCCESSFUL in 3s
5 actionable tasks: 5 executed
```

**Comment by grv87** [comment]
IIUC, the problem persisted with all versions of Gradle. It's just that recent versions perform validations more thoroughly.

**Comment by patton73** [comment]
These are not deprecations. But A sort of warnings. 

Gradle detected a problem with the following location: '...x.jar'. Reason: Task ':distTar' uses this output of task ':shadowJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.

Basically dependencies between internal shadowjar tasks are not declared. And gradle is slow because it does not know with task depends from another.

**Comment by johnrengelman** [comment]
I just don’t see these warnings when I run a project on Gradle 7.2 and shadow 7.1. What am I missing to get them?

**Comment by Angular-Angel** [comment]
Huh, so, I've been building my project with shadow for a while and I didn't have this problem - but somehow I did something and now I do have this problem, and I have no idea what i did. I didn't change my shadow version, it's still 7.1.2, nor did I change my gradle version - still 7.3.3. I don't know what the problem is. :/

**Comment by drew-corporate-creations** [comment]
I have a multi-project build with a number of 'library' projects and other projects that are dependent on them. Been a while since I upgraded anything. I just went from Gradle 6.2 to 7.4 and Shadow 5.0 to 7.1.2 and am getting these warnings now. The build is still successful.


`Execution optimizations have been disabled for task ':dynamo-database-module:shadowJar' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '../modules/common-micronaut-module/build/libs/common-micronaut-module-1.0-SNAPSHOT.jar'. Reason: Task ':dynamo-database-module:shadowJar' uses this output of task ':common-micronaut-module:shadowJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
`

**Comment by virustotalop** [comment]
Declaring this in your sub-module's gradle file where you depend on another sub-module seems to fix the issue `tasks.shadowJar.dependsOn ':subModuleYouDependOn:shadowJar'`. Using the [configuration shadow as seen here](https://imperceptiblethoughts.com/shadow/multi-project/), may also fix the issue but I haven't tried it because I ran into other issues with my configuration due to it.

**Comment by drew-corporate-creations** [comment]
> Declaring this in your module's gradle file where you depend on another module seems to fix the issue `tasks.taskThatDependsOnShadowJar.dependsOn ':moduleYouDependOn:shadowJar'`. For example `tasks.shadowJar.dependsOn ':common:shadowJar'`. Using the [configuration shadow as seen here](https://imperceptiblethoughts.com/shadow/multi-project/) may also fix the issue but I haven't tried it because I ran into other issues with my configuration due to it.

First solution works! Thank you. I found that same link you provided in the alternate solution as well, but it does not work.

**Comment by loicgreffier** [comment]
I don't know if this can be helpful, but I have run into the following similar warnings recently: 

> Task :shadowJar
Execution optimizations have been disabled for task ':shadowJar' to ensure correctness due to the following reasons:
>  - Gradle detected a problem with the following location: 'E:\Projects\kafkactl\build\libs\kafkactl-0.1.0-SNAPSHOT.jar'. Reason: Task ':distTar' uses this output of task ':shadowJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.5.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
>  - Gradle detected a problem with the following location: 'E:\Projects\kafkactl\build\libs\kafkactl-0.1.0-SNAPSHOT.jar'. Reason: Task ':distZip' uses this output of task ':shadowJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.5.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
>  - Gradle detected a problem with the following location: 'E:\Projects\kafkactl\build\libs\kafkactl-0.1.0-SNAPSHOT.jar'. Reason: Task ':startScripts' uses this output of task ':shadowJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.5.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.

**Gradle Version** 

7.5.1

**Shadow Version**

7.1.2

**Cause**

The issue came when I override the shadow jar archive name

```yml
shadowJar {
    archiveClassifier.set('')
}
```

**Solution**

Declaring the dependencies manually seems to solve the warnings

```yml
shadowJar {
    archiveClassifier.set('')
    dependsOn(distTar, distZip)
}
```

**Comment by schrek1** [comment]
thanks @loicgreffier works like charm

I add snippet for Kotlin DSL

```
tasks.withType<ShadowJar> {
    archiveClassifier.set("")
    dependsOn("distTar", "distZip")
} 
```

**Comment by Mattwmaster58** [comment]
@johnrengelman It doesn't seem like this is fixed, I'm seeing the issue with Gradle 8.14.2 and shadowJar 9.0.2 when I use `archiveClassifier.set('')`. The workarounds do work, but it would be nice to have this working out of the box, of course
